addons/portal/wizard/mail_invite_share.py

```diff
@@ -0,0 +1,51 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+import werkzeug.urls
+
+from odoo import api, fields, models, _
+
+
+class MailInviteShare(models.TransientModel):
+    _name = 'mail.invite.share'
+
+    def default_share_link(self):
+        active_model = self.env.context.get('active_model')
+        active_id = self.env.context.get('active_id')
+        res = False
+        if active_model:
+            model = self.env[active_model]
+            if isinstance(model, self.pool['portal.mixin']):
+                doc_url = model.browse(active_id).portal_url
+                base_url = self.env['ir.config_parameter'].sudo().get_param('web.base.url')
+                res = base_url + doc_url
+        return res
+
+    share_link = fields.Char(string="Document link", default=default_share_link)
+    partner_ids = fields.Many2many('res.partner', string="Recipients", required=True)
+    note = fields.Text(help="Add extra content to display in the email")
+
+    @api.multi
+    def send_mail_action(self):
+        active_model = self.env.context.get('active_model')
+        active_id = self.env.context.get('active_id')
+        active_record = self.env[active_model].browse(active_id)
+
+        base_url = self.env['ir.config_parameter'].sudo().get_param('web.base.url')
+        template = self.env.ref('portal.portal_share_document_template', False)
+        note = self.env.ref('mail.mt_note')
+
+        user_partner_ids = self.partner_ids.filtered(lambda x: x.user_ids)
+        # if partner already user send common link in batch to all user
+        if user_partner_ids:
+            query = dict(redirect=active_record.portal_url, db=self.env.cr.dbname)
```

Comment:
There is a mismatch to clean with portal_url, get_share_url and you custom share_link computation that seems to change.

Reply:
@tde-banana-odoo , 
As per spec 'user must log in to see the shared document' 
and if we use  get_share_url it will allow to access the document using access token(without login).
(because it generates token if the access_token field exists in the current model and also generate signup for partner_id if partner_id exist in the current model.)

So can we modify the behavior of get_share_url to support any partner and allow access token based on the argument?

'share_link' link is just full URL for the current record's portal_url(base_url + portal_url). 
Thanks